### DISCUSSION TITLE
example of openBLAS speeding up models?

### DISCUSSION THREAD

**Discussion by James-Thorson-NOAA** [sdmTMB/sdmTMB#373] [Q&A]
Hi all,

I can't seem to get an improvement in sdmTMB (or other TMB models) using the openBLAS instructions that are very thoroughly and helpfully provided.  Below, I show that the matrix-multiplication is faster (as expected) but Pcod spatial model is slower using the updated files relative to base versions.  

Does anyone have additional experience to share regarding enabling openBLAS on a windows machine using NOAA IT privileges?

Using the standard `Rblas.dll` and `Rlapack.dll` files:
![image](https://github.com/user-attachments/assets/a0e5fe21-40e4-4b2d-a6e1-f8478ad203fd)

and using the updated versions:
![image](https://github.com/user-attachments/assets/c8c1bcb5-d794-423c-9077-e71bcf7c8f74)

**Comment by seananderson** [comment]
Have you ruled out that openBLAS is using all available cores? That can slow things down. There's a bit at the bottom of [the Gist](https://gist.github.com/seananderson/08a51e296a854f227a908ddd365fb9c1) about that:

> You may find that openBLAS uses all available cores, which may not be computationally efficient with TMB. If so, you can specify BLAS should only use 1 core with the [RhpcBLASctl](https://cran.r-project.org/package=RhpcBLASctl) package:

```r
RhpcBLASctl::blas_set_num_threads(1)
RhpcBLASctl::omp_set_num_threads(1)
```

> You can either set that at the beginning of each script or (better yet) place it in your .Rprofile using:

```r
usethis::edit_r_profile()
```

Assuming the above is the default behaviour, maybe we should make those instructions more prominent or try to detect that for the user.

Apple's vecLib seems to default to 1 core.

Also, my memory is the speed up is more noticeable with an AR1 or random walk spatiotemporal field, but at the very least it shouldn't be slower for a simple spatial model!

**Comment by James-Thorson-NOAA** [comment]
Below is with the spatio-temporal RW turned on.  I've tried:

- playing with `RhpcBLASctl::blas_set_num_threads(24)` and `RhpcBLASctl::omp_set_num_threads(24)` vs. `RhpcBLASctl::blas_set_num_threads(1)` and `RhpcBLASctl::omp_set_num_threads(1)` and don't see a big change in CPU use in the "Task manager" ... I don't know if that's a sign that something is wrong?
- Re-installation `sdmTMB` locally in case I need to recompile the CPP for the changes to take effect
- Setting `OPENBLAS_NUM_THREADS` as environment variable as per [David-Cortes advice](https://github.com/david-cortes/R-openblas-in-windows?tab=readme-ov-file#instructions-in-detail)
... and none of these have seemed to fix things.

Of course it's not a huge issue (I can always use MRAN locally), but I'd be happy to hear if there's other intermediate diagnostics that might illustrate where the issue is arising.

With updated BLAS:
![image](https://github.com/user-attachments/assets/92683a70-9461-4513-835d-c83505957d5c)

and default BLAS:
![image](https://github.com/user-attachments/assets/c80e76b5-6720-4629-a1d3-bac77b7d0777)

**Comment by seananderson** [comment]
Hmm... I'm not sure. I'll have to try this myself on a DFO Windows laptop.

**Comment by tom-peatman** [comment]
Hi @James-Thorson-NOAA , @seananderson - apologies for jumping in on this thread. It's the first time that I've come across others having a similar experience to me (and with info on computational times that I can compare to mine).

I've been exploring using openBLAS to reduce fitting times for sdmTMB model. I'm fitting models to relatively large fisheries observer datasets (c. 250,000 observed sets), and we're interested in using cross validation ad model stacking as part of the workflow, so speeding up model fitting time is of interest.

I'm getting substantial reductions in computation time for matrix multiplication using openBLAS (0.3.30), but very modest reductions (or no reduction) in computation times when fitting sdmTMB models using openBLAS.

I've run the code chunks at the end of this post with R's default `Rblas.dll` and `Rlapack.dll` files, as well as the openBLAS (0.3.30) files, with the environment variable `OPENBLAS_NUM_THREADS` set to 1.

For the matrix multiplication, I'm getting a mean time of 2 seconds with R's default `Rblas.dll` and `Rlapack.dll` files, reducing to 140 milliseconds when using openBLAS. These look pretty consistent with expected reductions based on David Cortes' very helpful [page](https://github.com/david-cortes/R-openblas-in-windows), and @James-Thorson-NOAA's examples.

For the `sdmTMB` example, I'm getting a mean of 2.6 seconds with R's default `Rblas.dll` and `Rlapack.dll` files, reducing to 2.2 seconds when using openBLAS. That said, I haven't noticed any substantial reductions in model fitting times when I've using sdmTMB on the datasets I'm working on.

I'm using R version 4.4.1, and sdmTMB version 0.7.4.

I'd be interested to know whether either of you made any progress in determining what (if anything) was preventing speeding up model fitting with openBLAS, and would also be very grateful for any information or advice you can provide. Perhaps my expectations on reductions in runtimes are the issue... I think I'm running the same example as @James-Thorson-NOAA above, but my run times with the default `dll` files are already orders of magnitude faster.

Thanks in advance!

Fitting `sdmTMB` models:
```
library(microbenchmark)
library(sdmTMB)

# Build a mesh to implement the SPDE approach:
mesh <- make_mesh(pcod_2011, c("X", "Y"), cutoff = 20)

microbenchmark({
  sdmTMB(
    density ~ s(depth),
    data = pcod_2011, mesh = mesh,
    time = "year",
    spatiotemporal = "rw",
    family= tweedie(link = "log")
  )
}, times = 3L)
```

Matrix multiplication:
```
library(microbenchmark)
set.seed(1)
m <- 1e4
n <- 1e3
k <- 3e2
X <- matrix(rnorm(m*k), nrow=m)
Y <- matrix(rnorm(n*k), ncol=n)
microbenchmark({
    Z <- X %*% Y
}, times=10L)
```

**Comment by seananderson** [comment]
OpenBLAS primarily accelerates large dense linear algebra operations. This sdmTMB example may not be dominated by those computations, so switching BLAS libraries doesn’t make much difference. I don't get much of an improvement with the example above either. Where it makes a big difference is with models with many more random effects and less sparse precision matrices. E.g.:

```r
library(sdmTMB)
mesh <- make_mesh(pcod, c("X", "Y"), cutoff = 10)
tictoc::tic()
fit <- sdmTMB(
  density ~ s(depth),
  data = pcod,
  mesh = mesh,
  time = "year",
  spatiotemporal = "rw",
  family = delta_gamma(),
  silent = FALSE
)
tictoc::toc()

# Default BLAS:
# 91.967 sec elapsed

# Apple's vecLib:
# 15.427 sec elapsed

set.seed(1)
m <- 1e4; n <- 1e3; k <- 3e2
X <- matrix(rnorm(m*k), nrow=m); Y <- matrix(rnorm(n*k), ncol=n)
system.time(X %*% Y)

# Default BLAS:
# user  system elapsed
# 1.232   0.025   1.257

# Apple's vecLib:
# user  system elapsed
# 0.024   0.010   0.018
```

Also, just for reference for anyone stumbling on this thread, installation instructions for optimized BLAS options are summarized in the [README file](https://github.com/pbs-assess/sdmTMB?tab=readme-ov-file#installation) near the top.

**Comment by tom-peatman** [comment]
Thanks for taking the time to look at this Sean - much appreciated.

It sounds like the models / datasets that I've been working on did not have many of the computations that you describe, so openBLAS did not result in substantial increases in speed. However, I expect that may change as we start looking at more complex model structures.

Thanks again for your input on this, and for the `sdmTMB` package in general, it's fantastic to work with!
